### PR TITLE
avoid costly iterator for session property cache

### PR DESCRIPTION
**Pull request by njr-11**
Avoid iterating through all values of the session property cache when we only need to update a subset of the properties.  This iteration is costly because it aggressively loads all of the cached values.

### CODE REVIEW

**Comment by njr-11** [comment]
#build

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java [hunk 1]
```diff
@@ -205,7 +213,7 @@ protected Object loadOneValue(String attrName, BackedSession sess) {
             byte[] bytes = cacheStoreService.getCache(appName).get(key);
 
             if (trace && tc.isDebugEnabled())
-                Tr.debug(this, tc, key.toString(), bytes.length);
+                Tr.debug(this, tc, key.toString(), bytes == null ? null : bytes.length);
```
Comment (aguibert) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java]: perhaps add a message here indicating what the trace value is?  Something like:
`"byte length for session=" + key.toString()`

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java [hunk 2]
```diff
@@ -152,67 +157,65 @@ protected Object getValue(String id, BackedSession s) {
...
-                            doWrite = true;
                             if (trace && tc.isDebugEnabled()) {
                                 Tr.debug(this, tc, "doing app changes for " + id + " on thread " + t);
                             }
-                            if (sht != null) {
-                                vEnum = sht.keys();
-                            } else {
-                                vEnum = (new Hashtable()).keys();
-                            }
+                            propsToWrite = sht.keySet();
```
Comment (aguibert) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java]: `sht` is an unfortunate variable name...  Change to `syncHt`?

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java [hunk 3]
```diff
@@ -311,6 +308,23 @@ boolean handlePropertyHits(BackedSession d2) {
...
+                    oldValue = cacheStoreService.cache.get(sessionKey);
+                    if (oldValue == null)
+                        throw new UnsupportedOperationException(); // TODO implement code path where cache entry for session is expired. Delete the property entries?
+                    SessionInfo sessionInfo = new SessionInfo(oldValue).clone();
+                    if (propsToWrite != null)
+                        sessionInfo.addSessionPropertyIds(propsToWrite);
+                    if (propsToRemove != null)
+                        sessionInfo.removeSessionPropertyIds(propsToRemove);
+                    newValue = sessionInfo.getArrayList();
+                } while (!cacheStoreService.cache.replace(sessionKey, oldValue, newValue));
```
Comment (aguibert) [inline dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java]: I'm worried about two clients updating the same cache and looping indefinitely here.  We should write a test that tries to force this situation.  Also, we should implement a max-retry count here where we blow up after something like 3 retries, in order to sniff out whether or not this situation actually occurs.

File: dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java (hunk 2 above)
Comment (njr-11) [reply dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java]: I don't like the name either.  Currently it matches DatabaseHashMapMR from which this code was copied.  At this point, the code has diverged enough it should be okay to change it

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
#build

**Comment by njr-11** [comment]
Build has 3 test failures which are in unrelated componentry.

**Pull request merged**